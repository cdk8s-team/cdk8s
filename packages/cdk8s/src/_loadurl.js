// this is an executable program that downloads a file from a url (with
// redirects) and prints it's output to STDOUT. it is implemented as a program
// because node.js does not support issuing network requests synchronously and
// in the cdk we need everything to be synchronous (it all happens in the
// ctors). so we utilize `spawnSync` to spawn this program as a child process.
// alternatively we could have use `curl` but this is more portable.

const { http, https } = require('follow-redirects');
const { parse } = require('url');
const fs = require('fs');

const url = process.argv[2];
if (!url) {
  console.error(`Usage: ${process.argv[1]} URL`);
  process.exit(1);
}

const purl = parse(url);

if (!purl.protocol) {
  if (!purl.pathname) {
    throw new Error(`unable to parse pathname from file url: ${url}`);
  }

  process.stdout.write(fs.readFileSync(purl.pathname, 'utf-8'));
  process.exit(0);
}

const client = getHttpClient();

const get = client.get(url, response => {
  if (response.statusCode !== 200) {
    throw new Error(`${response.statusCode} response from http get: ${response.statusMessage}`);
  }

  response.on('data', chunk => process.stdout.write(chunk));
});

get.once('error', err => {
  throw new Error(`http error: ${err.message}`);
});

function getHttpClient() {
  switch (purl.protocol) {
    case 'http:': return http;
    case 'https:': return https;
    default:
      throw new Error(`unsupported protocol "${purl.protocol}" in url: ${url}`);
  }
}
