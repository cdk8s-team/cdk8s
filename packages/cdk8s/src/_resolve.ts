import { Lazy } from './lazy';

export function resolve(value: any): any {

  if (value == null) {
    return value;
  }

  // cdk8s token
  if (value instanceof Lazy) {
    const resolved = value.produce();
    return resolve(resolved);
  }

  // implicit token as generated by the cdk8s-cli (e.g IntOrString)
  if (typeof(value.resolve) === 'function') {
    const resolved = value.resolve();
    return resolve(resolved);
  }

  if (typeof(value) !== 'object') {
    return value;
  }

  if (Array.isArray(value)) {
    return value.map(x => resolve(x));
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    result[k] = resolve(v);
  }

  return result;

}
