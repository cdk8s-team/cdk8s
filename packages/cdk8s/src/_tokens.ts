import { Tokenization, DefaultTokenResolver, StringConcat, Construct } from 'constructs';

const TOKEN_RESOLVER = new DefaultTokenResolver(new StringConcat());

export function resolve<T>(scope: Construct, obj: T): T {
  return Tokenization.resolve(obj, {
    scope,
    resolver: TOKEN_RESOLVER,
  });
}