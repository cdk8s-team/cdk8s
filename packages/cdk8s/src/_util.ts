export interface SanitizeOptions {
  /**
   * Do not include empty objects (no keys).
   */
  readonly filterEmptyObjects?: boolean;

  /**
   * Do not include arrays with no items.
   */
  readonly filterEmptyArrays?: boolean;
}

export function sanitizeValue(obj: any, options: SanitizeOptions = { }): any {
  if (obj == null) {
    return undefined;
  }

  if (typeof(obj) !== 'object') {
    return obj;
  }

  if (Array.isArray(obj)) {

    if (options.filterEmptyArrays && obj.length === 0) {
      return undefined;
    }

    return obj.map(x => sanitizeValue(x, options));
  }

  if (obj.constructor.name !== 'Object') {
    throw new Error(`can't render non-simple object of type '${obj.constructor.name}'`);
  }

  const newObj: { [key: string]: any } = { };

  for (const [key, value] of Object.entries(obj)) {
    const newValue = sanitizeValue(value, options);
    if (newValue != null) {
      newObj[key] = newValue;
    }
  }

  if (options.filterEmptyObjects && Object.keys(newObj).length === 0) {
    return undefined;
  }

  return newObj;
}
