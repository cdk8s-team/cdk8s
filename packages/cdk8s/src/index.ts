export * from './api-object';
export * from './chart';
export * from './testing';
export * from './app';
export * from './include';
export * from './yaml';
export * from './metadata';