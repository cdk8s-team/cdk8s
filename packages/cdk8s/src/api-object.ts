import { Construct, IConstruct, Node } from 'constructs';
import { Chart } from './chart';
import { sanitizeValue } from './_util';
import { resolve } from './_tokens';
import * as stringify from 'json-stable-stringify';
import { ApiObjectMetadata, ApiObjectMetadataDefinition } from './metadata';

/**
 * Options for defining API objects.
 */
export interface ApiObjectOptions {
  /**
   * Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ApiObjectMetadata;

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * Additional attributes for this API object.
   */
  readonly [key: string]: any;
}

export class ApiObject extends Construct {
  /**
   * The name of the API object.
   *
   * If a name is specified in `metadata.name` this will be the name returned.
   * Otherwise, a name will be generated by calling
   * `Chart.of(this).generatedObjectName(this)`, which by default uses the
   * construct path to generate a DNS-compatible name for the resource.
   */
  public readonly name: string;

  /**
   * The object's API version.
   */
  public readonly apiVersion: string;

  /**
   * The object kind.
   */
  public readonly kind: string;

  /**
   * The chart in which this object is defined.
   */
  public readonly chart: Chart;

  /**
   * Metadata associated with this API object.
   */
  public readonly metadata: ApiObjectMetadataDefinition;

  /**
   * Defines an API object.
   *
   * @param scope the construct scope
   * @param ns namespace
   * @param options options
   */
  constructor(scope: Construct, ns: string, private readonly options: ApiObjectOptions) {
    super(scope, ns);
    this.chart = Chart.of(this);
    this.kind = options.kind;
    this.apiVersion = options.apiVersion;

    this.name = options.metadata?.name ?? this.chart.generateObjectName(this);

    this.metadata = new ApiObjectMetadataDefinition({
      name: this.name,
      namespace: options.metadata?.namespace ?? this.chart.namespace,

      // override user defined values
      ...options.metadata,
    });
  }

  /**
   * Create a dependency between this ApiObject and other constructs.
   * These can be other ApiObjects, Charts, or custom.
   *
   * @param dependencies the dependencies to add.
   */
  public addDependency(...dependencies: IConstruct[]) {
    Node.of(this).addDependency(...dependencies);
  }

  /**
   * Renders the object to Kubernetes JSON.
   */
  public toJson(): any {
    const data = {
      ...this.options,
      metadata: this.metadata.toJson(),
    };

    // convert to "pure data" so, for example, when we convert to yaml these
    // references are not converted to anchors.
    return JSON.parse(stringify(sanitizeValue(resolve(this, data))));
  }
}
