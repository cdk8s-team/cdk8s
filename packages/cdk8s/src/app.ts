import * as fs from 'fs';
import * as path from 'path';
import { Construct, Node, IConstruct } from 'constructs';
import { ApiObject } from './api-object';
import { Chart } from './chart';
import { DependencyGraph } from './dependency';
import { Names } from './names';
import { Yaml } from './yaml';

export interface AppProps {
  /**
   * The directory to output Kubernetes manifests.
   *
   * @default - CDK8S_OUTDIR if defined, otherwise "dist"
   */
  readonly outdir?: string;
}

/**
 * Represents a cdk8s application.
 */
export class App extends Construct {
  /**
   * Synthesize a single chart.
   *
   * Each element returned in the resulting array represents a different ApiObject
   * in the scope of the chart.
   *
   * Note that the returned array order is important. It is determined by the various dependencies between
   * the constructs in the chart, where the first element is the one without dependencies, and so on...
   *
   * @returns An array of JSON objects.
   * @param chart the chart to synthesize.
   * @internal
   */
  public static _synthChart(chart: Chart): any[] {

    const app: App = App.of(chart);

    // we must prepare the entire app before synthesizing the chart
    // because the dependency inference happens on the app level.
    resolveDependencies(app);

    // validate the app since we want to call onValidate of the relevant constructs.
    // note this will also call onValidate on constructs from possibly different charts,
    // but thats ok too since we no longer treat constructs as a self-contained synthesis unit.
    validate(app);

    return chartToKube(chart);
  }

  private static of(c: IConstruct): App {

    const scope = Node.of(c).scope;

    if (!scope) {
      // the app is the only construct without a scope.
      return c as App;
    }

    return App.of(scope);
  }

  /**
   * The output directory into which manifests will be synthesized.
   */
  public readonly outdir: string;

  /**
   * Defines an app
   * @param props configuration options
   */
  constructor(props: AppProps = { }) {
    super(undefined as any, '');
    this.outdir = props.outdir ?? process.env.CDK8S_OUTDIR ?? 'dist';
  }

  /**
   * Synthesizes all manifests to the output directory
   */
  public synth(): void {

    fs.mkdirSync(this.outdir, { recursive: true });

    // this is kind of sucky, eventually I would like the DependencyGraph
    // to be able to answer this question.
    const hasDependantCharts = resolveDependencies(this);

    // Since we plan on removing the distributed synth mechanism, we no longer call `Node.synthesize`, but rather simply implement
    // the necessary operations. We do however want to preserve the distributed validation.
    validate(this);

    const simpleManifestNamer = (chart: Chart) => `${Names.toDnsLabel(Node.of(chart).path)}.k8s.yaml`;
    const manifestNamer = hasDependantCharts ? (chart: Chart) => `${index.toString().padStart(4, '0')}-${simpleManifestNamer(chart)}` : simpleManifestNamer;

    const charts: IConstruct[] = new DependencyGraph(Node.of(this)).topology().filter(x => x instanceof Chart);

    let index = 0;
    for (const node of charts) {
      const chart: Chart = Chart.of(node);
      Yaml.save(path.join(this.outdir, manifestNamer(chart)), chartToKube(chart));
      index++;
    }

  }
}

function validate(app: App) {

  // Note this is a copy-paste of https://github.com/aws/constructs/blob/master/lib/construct.ts#L438.
  const errors = Node.of(app).validate();
  if (errors.length > 0) {
    const errorList = errors.map(e => `[${Node.of(e.source).path}] ${e.message}`).join('\n  ');
    throw new Error(`Validation failed with the following errors:\n  ${errorList}`);
  }

}

function resolveDependencies(app: App) {

  let hasDependantCharts = false;

  for (const dep of Node.of(app).dependencies) {

    // create explicit api object dependencies from implicit construct dependencies
    const targetApiObjects = Node.of(dep.target).findAll().filter(c => c instanceof ApiObject);
    const sourceApiObjects = Node.of(dep.source).findAll().filter(c => c instanceof ApiObject);

    for (const target of targetApiObjects) {
      for (const source of sourceApiObjects) {
        Node.of(source).addDependency(target);
      }
    }

    // create an explicit chart dependency from implicit construct dependencies
    const sourceChart = Chart.of(dep.source);
    const targetChart = Chart.of(dep.target);

    if (sourceChart !== targetChart) {
      Node.of(sourceChart).addDependency(targetChart);
      hasDependantCharts = true;
    }

  }

  return hasDependantCharts;

}

function chartToKube(chart: Chart) {
  return new DependencyGraph(Node.of(chart)).topology()
    .filter(x => x instanceof ApiObject)
    .map(x => (x as ApiObject).toJson());
}
