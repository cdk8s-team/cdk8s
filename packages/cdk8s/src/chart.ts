import { Construct, Node, IConstruct } from 'constructs';
import { ApiObject } from './api-object';
import { App } from './app';
import { Names } from './names';

export interface ChartProps {
  /**
   * The default namespace for all objects defined in this chart (directly or
   * indirectly). This namespace will only apply to objects that don't have a
   * `namespace` explicitly defined for them.
   *
   * @default - no namespace is synthesized (usually this implies "default")
   */
  readonly namespace?: string;

  /**
   * Labels to apply to all resources in this chart.
   *
   * @default - no common labels
   */
  readonly labels?: { [name: string]: string };
}

export class Chart extends Construct {

  /**
   * Finds the chart in which a node is defined.
   * @param c a construct node
   */
  public static of(c: IConstruct): Chart {
    if (c instanceof Chart) {
      return c;
    }

    const parent = Node.of(c).scope as Construct;
    if (!parent) {
      throw new Error('cannot find a parent chart (directly or indirectly)');
    }

    return Chart.of(parent);
  }

  /**
   * The default namespace for all objects in this chart.
   */
  public readonly namespace?: string;

  /**
   * Chart-level labels.
   */
  private readonly _labels?: { [name: string]: string };

  constructor(scope: Construct, id: string, props: ChartProps = { }) {
    super(scope, id);
    this.namespace = props.namespace;
    this._labels = props.labels ?? {};
  }

  /**
   * Labels applied to all resources in this chart.
   *
   * This is an immutable copy.
   */
  public get labels(): { [name: string]: string } {
    return { ...this._labels };
  }

  /**
   * Generates a app-unique name for an object given it's construct node path.
   *
   * Different resource types may have different constraints on names
   * (`metadata.name`). The previous version of the name generator was
   * compatible with DNS_SUBDOMAIN but not with DNS_LABEL.
   *
   * For example, `Deployment` names must comply with DNS_SUBDOMAIN while
   * `Service` names must comply with DNS_LABEL.
   *
   * Since there is no formal specification for this, the default name
   * generation scheme for kubernetes objects in cdk8s was changed to DNS_LABEL,
   * since itâ€™s the common denominator for all kubernetes resources
   * (supposedly).
   *
   * You can override this method if you wish to customize object names at the
   * chart level.
   *
   * @param apiObject The API object to generate a name for.
   */
  public generateObjectName(apiObject: ApiObject) {
    return Names.toDnsLabel(apiObject);
  }

  /**
   * Create a dependency between this Chart and other constructs.
   * These can be other ApiObjects, Charts, or custom.
   *
   * @param dependencies the dependencies to add.
   */
  public addDependency(...dependencies: IConstruct[]) {
    Node.of(this).addDependency(...dependencies);
  }

  /**
   * Renders this chart to a set of Kubernetes JSON resources.
   * @returns array of resource manifests
   */
  public toJson(): any[] {
    return App._synthChart(this);
  }
}

