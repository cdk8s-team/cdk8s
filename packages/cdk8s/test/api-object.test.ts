import { Construct, Node, Dependency } from 'constructs';
import { ApiObject, Chart, JsonPatch, Testing } from '../src';

test('minimal configuration', () => {
  const app = Testing.app();
  const stack = new Chart(app, 'test');

  new ApiObject(stack, 'my-resource', {
    apiVersion: 'v1',
    kind: 'MyResource',
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('printed yaml is alphabetical', () => {
  const app = Testing.app();
  const stack = new Chart(app, 'test');

  // Object keys in random order
  new ApiObject(stack, 'my-resource', {
    kind: 'MyResource',
    apiVersion: 'v1',
    spec: {
      secondProperty: {
        innerThirdProperty: '!',
        beforeThirdProperty: 'world',
      },
      firstProperty: 'hello',
    },
  });

  // Should match alphabetically-ordered snapshot
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('addDependency', () => {

  const app = Testing.app();
  const chart = new Chart(app, 'chart1');

  const obj1 = new ApiObject(chart, 'obj1', { apiVersion: 'v1', kind: 'Kind1' });
  const obj2 = new ApiObject(chart, 'obj2', { apiVersion: 'v1', kind: 'Kind2' });
  const obj3 = new ApiObject(chart, 'obj3', { apiVersion: 'v1', kind: 'Kind3' });

  obj1.addDependency(obj2, obj3);

  const dependencies: Set<Dependency> = new Set<Dependency>(Node.of(obj1).dependencies);

  expect(dependencies).toEqual(new Set<Dependency>([
    {
      source: obj1,
      target: obj2,
    },
    {
      source: obj1,
      target: obj3,
    },
  ]));

});

test('synthesized resource name is based on path', () => {
  // GIVEN
  const app = Testing.app();
  const stack = new Chart(app, 'test');
  new ApiObject(stack, 'my-resource', {
    apiVersion: 'v1',
    kind: 'MyResource',
  });

  // WHEN
  // now create another resource with the same namespace but within a sub-scope
  const scope = new Construct(stack, 'scope');
  new ApiObject(scope, 'my-resource', {
    apiVersion: 'v1',
    kind: 'MyResource',
  });

  // THEN
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('if name is explicitly specified it will be respected', () => {
  // GIVEN
  const app = Testing.app();
  const stack = new Chart(app, 'test');

  // WHEN
  new ApiObject(stack, 'my-resource', {
    kind: 'MyResource',
    apiVersion: 'v1',
    metadata: {
      name: 'boom',
    },
  });

  // THEN
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('"spec" is synthesized as-is', () => {
  // GIVEN
  const app = Testing.app();
  const stack = new Chart(app, 'test');

  // WHEN
  new ApiObject(stack, 'resource', {
    kind: 'ResourceType',
    apiVersion: 'v1',
    spec: {
      prop1: 'hello',
      prop2: {
        world: 123,
      },
    },
  });

  // THEN
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('"data" can be used to specify resource data', () => {
  // GIVEN
  const app = Testing.app();
  const stack = new Chart(app, 'test');

  // WHEN
  new ApiObject(stack, 'resource', {
    kind: 'ResourceType',
    apiVersion: 'v1',
    data: {
      boom: 123,
    },
  });

  // THEN
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('object naming logic can be overridden at the chart level', () => {
  class MyChart extends Chart {
    generateObjectName() {
      return 'fixed!';
    }
  }

  // GIVEN
  const app = Testing.app();
  const chart = new MyChart(app, 'my-chart');

  // WHEN
  const object = new ApiObject(chart, 'my-object', {
    apiVersion: 'v1',
    kind: 'MyKind',
  });

  // THEN
  expect(object.name).toEqual('fixed!');
  expect(Testing.synth(chart)).toStrictEqual([{
    apiVersion: 'v1',
    kind: 'MyKind',
    metadata: { name: 'fixed!' },
  }]);
});

test('default namespace can be defined at the chart level', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'chart', { namespace: 'ns1' });
  const group1 = new Construct(chart, 'group1');

  // WHEN
  new ApiObject(group1, 'obj1', { apiVersion: 'v1', kind: 'Kind1' });
  new ApiObject(group1, 'obj2', { apiVersion: 'v2', kind: 'Kind2', metadata: { namespace: 'foobar' } });

  // THEN
  expect(Testing.synth(chart)).toStrictEqual([{
    apiVersion: 'v1',
    kind:
    'Kind1',
    metadata: {
      name: 'chart-group1-obj1-b4b7a9d0',
      namespace: 'ns1',
    },
  }, {
    apiVersion: 'v2',
    kind: 'Kind2',
    metadata: {
      name: 'chart-group1-obj2-3cdc9d22',
      namespace: 'foobar',
    },
  }]);
});

test('chart labels are applied to all api objects in the chart', () => {
  // GIVEN
  const app = Testing.app();

  // WHEN
  const chart = new Chart(app, 'my-chart', {
    labels: {
      foo: 'ffffffffff',
      bar: 'bbbbbb',
    },
  });

  new ApiObject(chart, 'obj1', { kind: 'Obj1', apiVersion: 'v1' });
  const group = new Construct(chart, 'group');
  new ApiObject(group, 'obj2', {
    kind: 'Obj2',
    apiVersion: 'v2',
    metadata: {
      labels: {
        foo: 'override by object',
        zoo: 'zoo1',
      },
    },
  });

  // THEN
  expect(Testing.synth(chart)).toEqual([
    {
      apiVersion: 'v1',
      kind: 'Obj1',
      metadata: {
        labels: {
          bar: 'bbbbbb',
          foo: 'ffffffffff',
        },
        name: 'my-chart-obj1-01c0df67',
      },
    },
    {
      apiVersion: 'v2',
      kind: 'Obj2',
      metadata: {
        labels: {
          bar: 'bbbbbb',
          foo: 'override by object',
          zoo: 'zoo1',
        },
        name: 'my-chart-group-obj2-ba1eb578',
      },
    },
  ]);
});

describe('ApiObject.of()', () => {

  test('fails if there is no default child', () => {
    // GIVEN
    const chart = Testing.chart();

    // THEN
    expect(() => ApiObject.of(new Construct(chart, 'hello'))).toThrow(/cannot find a \(direct or indirect\) child of type ApiObject/);
  });

  test('returns the object if it is an API object', () => {
    // GIVEN
    const chart = Testing.chart();
    const obj = new ApiObject(chart, 'my-obj', { apiVersion: 'v1', kind: 'Foo' });

    // THEN
    expect(ApiObject.of(obj)).toBe(obj);
  });

  test('returns a direct child', () => {
    // GIVEN
    const chart = Testing.chart();
    const parent = new Construct(chart, 'l2');

    // WHEN
    const obj = new ApiObject(parent, 'Default', { apiVersion: 'v1', kind: 'Foo' });

    // THEN
    expect(ApiObject.of(parent)).toBe(obj);
  });

  test('returns an indirect child', () => {
    // GIVEN
    const chart = Testing.chart();
    const parent1 = new Construct(chart, 'l3');

    // WHEN
    const parent2 = new Construct(parent1, 'Default');
    const obj = new ApiObject(parent2, 'Default', { apiVersion: 'v1', kind: 'Foo' });

    // THEN
    expect(ApiObject.of(parent1)).toBe(obj);
  });

});

describe('addJsonPatch()', () => {

  test('applied after the object has been synthesized', () => {
    // GIVEN
    const chart = Testing.chart();
    const obj = new ApiObject(chart, 'obj', {
      kind: 'Obj',
      apiVersion: 'v1',
      spec: {
        foo: 1234,
        bar: {
          baz: [1, 2, 3],
        },
      },
    });

    // WHEN
    obj.addJsonPatch(JsonPatch.add('/spec/bar/baz/3', 4));
    obj.addJsonPatch(JsonPatch.remove('/spec/foo'));
    obj.addJsonPatch(JsonPatch.copy('/apiVersion', '/spec/apiVersion'));

    // THEN
    expect(Testing.synth(chart)).toStrictEqual([
      {
        apiVersion: 'v1',
        kind: 'Obj',
        metadata: {
          name: 'test-obj-822728f1',
        },
        spec: {
          apiVersion: 'v1',
          bar: {
            baz: [1, 2, 3, 4],
          },
        },
      },
    ]);
  });

});
