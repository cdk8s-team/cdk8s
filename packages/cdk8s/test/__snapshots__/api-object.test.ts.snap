// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"data" can be used to specify resource data 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "boom": 123,
    },
    "kind": "ResourceType",
    "metadata": Object {
      "name": "test-c8c5facf",
    },
  },
]
`;

exports[`"spec" is synthesized as-is 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "ResourceType",
    "metadata": Object {
      "name": "test-c8c5facf",
    },
    "spec": Object {
      "prop1": "hello",
      "prop2": Object {
        "world": 123,
      },
    },
  },
]
`;

exports[`if name is explicitly specified it will be respected 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "MyResource",
    "metadata": Object {
      "name": "boom",
    },
  },
]
`;

exports[`minimal configuration 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "MyResource",
    "metadata": Object {
      "name": "test-my-c8487bf7",
    },
  },
]
`;

exports[`printed yaml is alphabetical 1`] = `
"[
  {
    \\"apiVersion\\": \\"v1\\",
    \\"kind\\": \\"MyResource\\",
    \\"metadata\\": {
      \\"meta\\": {
        \\"aaa\\": 123,
        \\"zzz\\": \\"hello\\"
      },
      \\"name\\": \\"test-my-c8487bf7\\"
    },
    \\"spec\\": {
      \\"firstProperty\\": \\"hello\\",
      \\"secondProperty\\": {
        \\"beforeThirdProperty\\": \\"world\\",
        \\"innerThirdProperty\\": \\"!\\"
      }
    }
  }
]"
`;

exports[`synthesized resource name is based on path 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "MyResource",
    "metadata": Object {
      "name": "test-my-c8487bf7",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "MyResource",
    "metadata": Object {
      "name": "test-scope-my-c8fafaf7",
    },
  },
]
`;
